libp2p/protocols/secure/noise.nim

```diff
@@ -509,21 +530,28 @@ method write*(
   # sequencing issues
   sconn.stream.write(cipherFrames)
 
-method handshake*(p: Noise, conn: Connection, initiator: bool, peerId: Opt[PeerId]): Future[SecureConn] {.async.} =
+method handshake*(
+    p: Noise,
+    conn: Connection,
+    initiator: bool,
+    peerId: Opt[PeerId]
+): Future[SecureConn] {.async: (raises: [CancelledError, LPStreamError]).} =
   trace "Starting Noise handshake", conn, initiator
 
   let timeout = conn.timeout
   conn.timeout = HandshakeTimeout
 
   # https://github.com/libp2p/specs/tree/master/noise#libp2p-data-in-handshake-messages
-  let
-    signedPayload = p.localPrivateKey.sign(
-      PayloadString & p.noiseKeys.publicKey.getBytes).tryGet()
+  let signedPayload = p.localPrivateKey.sign(
+    PayloadString & p.noiseKeys.publicKey.getBytes)
+  if signedPayload.isErr():
+    raise newException(NoiseHandshakeError,
+      "Failed to sign public key: " & $signedPayload.error())
```

Comment:
The `ResultError[T]` from `tryGet` is now wrapped into `NoiseHandshakeError`. Similarly for other `tryGet` in `secio`. Conceptually, it makes sense to restrict `handshake` to `LPStreamError` (and descendants).

Reply:
technically, there exists `toException` to make `tryGet` raise something else, but it's not a widely used mechanism.. https://github.com/arnetheduck/nim-results?tab=readme-ov-file#example

Reply:
Like, one of the motivations of annotating libp2p with `{.async: (raises).}` is to have a chance to get rid of `toException` converting every single `cstring/string` of whoever pulls in libp2p as a dependency. Once the annotations are done, `toException` can be replaced with explicit converter `mapErr(lpErr)` at the call sites that still need it. Doing it blindly is not possible though, as there are 400 callers and not all of them need it.

Reply:
It's triggered by @tersec request of me adding the `{.push raises: [].}` to test_keymanager_api so that he doesn't get hit by random linter error when bumping references / doing mechanical work on it. I noticed that in `test_keymanager_api` there are random functions that start requiring `{.raises: [LPError].}` because the implicit converter gets pulled in.

In `nimbus-eth2` we also have `tryGet` for downloading genesis states, but luckily they have a `except: CatchableError` that would also catch the `LPError` instead of `ResultError[cstring]`.

Reply:
https://status-im.github.io/nim-style-guide/language.converters.html

> Surprising conversions lead to ambiguous calls